# SIT720

# Machine Learning

# TASK 5.2C

In [1]:
import pandas as pd
import numpy as np
# Read the CSV file
data = pd.read_csv("cis.csv")

# Display the first few rows of the DataFrame
print("First few rows of the DataFrame:")
print(data.head())

missing_values = data.isnull().sum()

print("Null values in each column:")
print(missing_values)

# Assuming 'data' is your original DataFrame with missing values
# You can fill missing values with a specified value, such as 0
data_filled = data.fillna(0)

# Calculate and display the size of the dataset
size = data_filled.shape
print("\nSize of the dataset after dropping rows with missing categorical values:", size)


First few rows of the DataFrame:
   ID  N_Days Status             Drug    Age Sex Ascites Hepatomegaly Spiders  \
0   1     400      D  D-penicillamine  21464   F       Y            Y       Y   
1   2    4500      C  D-penicillamine  20617   F       N            Y       Y   
2   3    1012      D  D-penicillamine  25594   M       N            N       N   
3   4    1925      D  D-penicillamine  19994   F       N            Y       Y   
4   5    1504     CL          Placebo  13918   F       N            Y       Y   

  Edema  Bilirubin  Cholesterol  Albumin  Copper  Alk_Phos    SGOT  \
0     Y       14.5        261.0     2.60   156.0    1718.0  137.95   
1     N        1.1        302.0     4.14    54.0    7394.8  113.52   
2     S        1.4        176.0     3.48   210.0     516.0   96.10   
3     S        1.8        244.0     2.54    64.0    6121.8   60.63   
4     N        3.4        279.0     3.53   143.0     671.0  113.15   

   Tryglicerides  Platelets  Prothrombin  Stage  
0        

Reading CSV file: It imports the pandas library and reads a CSV file named "cirrhosis.csv" using the pd.read_csv() function. The data read from the CSV file is stored in a DataFrame named data.

Displaying the first few rows: It prints the first few rows of the DataFrame using the head() method, which provides a quick view of the data structure and contents.

Calculating null values: It calculates the number of missing values in each column of the DataFrame using the isnull() method to create a boolean DataFrame where missing values are marked as True, and then applies the sum() method to count the number of True values for each column.

Printing null values: It prints the number of null values in each column using the print() function.

Calculating dataset size: It calculates the size of the DataFrame after dropping rows with missing categorical values. However, the code snippet doesn't actually drop any rows or fill missing values, so the data_filled DataFrame referenced here seems to be missing in the provided code. Hence, this part of the code may raise an error. It calculates the size using the shape attribute, which returns a tuple containing the number of rows and columns in the DataFrame.

In [2]:
# Save the filled dataset to a new CSV file
data_filled.to_csv("cirrhosis_test.csv", index=False)

print("Filled dataset saved to cirrhosis_test.csv")

Filled dataset saved to cirrhosis_test.csv


In [3]:
# Extracting features and labels for training and testing
X_train = data_filled.iloc[:224, 3:]  
y_train = data_filled.iloc[:224, 2]   
X_test = data_filled.iloc[224:, 3:]   
y_test = data_filled.iloc[224:, 2]    

# Display training data information
print("Training Data Information:")
print("Number of samples in training data:", len(X_train))
print("Number of features in training data:", len(X_train.columns))

# Display missing values in training data
print("\nMissing values in Training Data:")
print(X_train.isnull().sum())

# Display test data information
print("\nTest Data Information:")
print("Number of samples in test data:", len(X_test))
print("Number of features in test data:", len(X_test.columns))

# Display missing values in test data
print("\nMissing values in Test Data:")
print(X_test.isnull().sum())


Training Data Information:
Number of samples in training data: 224
Number of features in training data: 17

Missing values in Training Data:
Drug             0
Age              0
Sex              0
Ascites          0
Hepatomegaly     0
Spiders          0
Edema            0
Bilirubin        0
Cholesterol      0
Albumin          0
Copper           0
Alk_Phos         0
SGOT             0
Tryglicerides    0
Platelets        0
Prothrombin      0
Stage            0
dtype: int64

Test Data Information:
Number of samples in test data: 88
Number of features in test data: 17

Missing values in Test Data:
Drug             0
Age              0
Sex              0
Ascites          0
Hepatomegaly     0
Spiders          0
Edema            0
Bilirubin        0
Cholesterol      0
Albumin          0
Copper           0
Alk_Phos         0
SGOT             0
Tryglicerides    0
Platelets        0
Prothrombin      0
Stage            0
dtype: int64


Extracting features and labels: It extracts features and labels for training and testing from the DataFrame data_filled. For training data, it selects the first 224 rows and all columns starting from the 4th column (index 3) to the last column. For testing data, it selects rows from index 224 to the end for both features and labels.

Displaying training data information: It prints information about the training data, including the number of samples (rows) and the number of features (columns) using the len() function and the shape attribute.

Displaying missing values in training data: It prints the number of missing values for each feature in the training data using the isnull() method followed by the sum() method to count the missing values for each column.

Displaying test data information: Similar to training data, it prints information about the test data, including the number of samples (rows) and the number of features (columns) using the len() function and the shape attribute.

Displaying missing values in test data: It prints the number of missing values for each feature in the test data using the isnull() method followed by the sum() method to count the missing values for each column.

In [4]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Find categorical columns
categorical_cols = X_train.select_dtypes(include=['object']).columns

# Impute missing values for numerical features with mean
numeric_transformer = SimpleImputer(strategy='mean')

# One-hot encode categorical features
categorical_transformer = OneHotEncoder()

# Combine transformers
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, ~X_train.columns.isin(categorical_cols)),
        ('cat', categorical_transformer, categorical_cols)
    ])

# Preprocess training data
X_train_imputed = pd.DataFrame(preprocessor.fit_transform(X_train), columns=preprocessor.get_feature_names_out(input_features=X_train.columns))

# Preprocess test data
X_test_imputed = pd.DataFrame(preprocessor.transform(X_test), columns=preprocessor.get_feature_names_out(input_features=X_test.columns))

# Display missing values after imputation
print("Missing values in Training Data after imputation:")
print(X_train_imputed.isnull().sum())
print("\nMissing values in Test Data after imputation:")
print(X_test_imputed.isnull().sum())


Missing values in Training Data after imputation:
num__Age                     0
num__Bilirubin               0
num__Cholesterol             0
num__Albumin                 0
num__Copper                  0
num__Alk_Phos                0
num__SGOT                    0
num__Tryglicerides           0
num__Platelets               0
num__Prothrombin             0
num__Stage                   0
cat__Drug_D-penicillamine    0
cat__Drug_Placebo            0
cat__Sex_F                   0
cat__Sex_M                   0
cat__Ascites_N               0
cat__Ascites_Y               0
cat__Hepatomegaly_N          0
cat__Hepatomegaly_Y          0
cat__Spiders_N               0
cat__Spiders_Y               0
cat__Edema_N                 0
cat__Edema_S                 0
cat__Edema_Y                 0
dtype: int64

Missing values in Test Data after imputation:
num__Age                     0
num__Bilirubin               0
num__Cholesterol             0
num__Albumin                 0
num__Copper           

Finding categorical columns: It identifies the categorical columns in the training data (X_train) using the select_dtypes method with the include=['object'] parameter, which selects columns of object data type.

Imputing missing values for numerical features: It initializes a SimpleImputer object to impute missing values for numerical features with the mean value. This imputer will replace missing values in numerical columns with the mean of the respective columns.

One-hot encoding categorical features: It initializes a OneHotEncoder object to perform one-hot encoding on categorical features. One-hot encoding converts categorical variables into binary vectors, which can be used as features in machine learning models.

Combining transformers: It creates a ColumnTransformer object named preprocessor to combine the imputation and one-hot encoding transformers. The ColumnTransformer applies different transformations to different columns of the input data.

Preprocessing training and test data: It preprocesses the training and test data using the fit_transform method of the ColumnTransformer for the training data and the transform method for the test data. This applies the imputation and one-hot encoding transformations to the respective datasets.



In [5]:
# Calculate label distribution
label_distribution = y_train.value_counts()

# Display label distribution
print("Label Distribution based on the Training Data:")
print(label_distribution)

# Check if the dataset is balanced
balance_threshold = 0.5
min_count = label_distribution.min()
max_count = label_distribution.max()
is_balanced = min_count / max_count >= balance_threshold

# Print result
if is_balanced:
    print("\nThe training set is balanced.")
else:
    print("\nThe training set is not balanced.")


Label Distribution based on the Training Data:
Status
D     114
C     103
CL      7
Name: count, dtype: int64

The training set is not balanced.


In [6]:
# Calculate label distribution for the test data
test_label_distribution = y_test.value_counts()

# Display label distribution for the test data
print("Label Distribution based on the Test Data:")
print(test_label_distribution)

# Check if the test dataset is balanced
test_min_count = test_label_distribution.min()
test_max_count = test_label_distribution.max()
test_is_balanced = test_min_count / test_max_count >= balance_threshold

# Print result for the test data
if test_is_balanced:
    print("\nThe test set is balanced.")
else:
    print("\nThe test set is not balanced.")


Label Distribution based on the Test Data:
Status
C     65
CL    12
D     11
Name: count, dtype: int64

The test set is not balanced.


In [7]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Create instances of the classifiers
models = {
    "Decision Tree": DecisionTreeClassifier(),
    "Random Forest": RandomForestClassifier(),
    "SVM": SVC(),
    "Logistic Regression": LogisticRegression(),
    "KNN": KNeighborsClassifier()
}

# Train and evaluate each model
for name, model in models.items():
    clf = make_pipeline(StandardScaler(), model)
    clf.fit(X_train_imputed, y_train)
    y_pred = clf.predict(X_test_imputed)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted', zero_division=1)
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    print(f"{name} Classifier:")
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"F1 Score: {f1:.4f}")


Decision Tree Classifier:
Accuracy: 0.6136
Precision: 0.7492
Recall: 0.6136
F1 Score: 0.6364
Random Forest Classifier:
Accuracy: 0.6023
Precision: 0.8884
Recall: 0.6023
F1 Score: 0.6236
SVM Classifier:
Accuracy: 0.6364
Precision: 0.8373
Recall: 0.6364
F1 Score: 0.6342
Logistic Regression Classifier:
Accuracy: 0.6591
Precision: 0.8422
Recall: 0.6591
F1 Score: 0.6519
KNN Classifier:
Accuracy: 0.6136
Precision: 0.8457
Recall: 0.6136
F1 Score: 0.6201


The dataset is first preprocessed to handle missing values and encode categorical variables. The following classifiers are used: Decision Tree, Random Forest, Support Vector Machine (SVM), Logistic Regression, and k-Nearest Neighbors (KNN).

Preprocessing:
Missing values in X_train and X_test are imputed using a column transformer that applies the mean strategy for numerical features and one-hot encoding for categorical features. This results in X_train_imputed and X_test_imputed.

Model Training and Evaluation:
Each model is instantiated and stored in a dictionary called models.
For each model, a pipeline is created that includes StandardScaler for feature scaling followed by the classifier.
The model is trained (fit) on the training data (X_train_imputed and y_train).
Predictions (y_pred) are made on the test data (X_test_imputed).

Metrics Calculation:
Accuracy, precision, recall, and F1 score are calculated for each model using the true labels (y_test) and the predictions (y_pred).
These metrics evaluate the model's performance, with the following interpretations:
Accuracy: The proportion of correctly predicted instances.
Precision: The proportion of true positive predictions among all positive predictions.
Recall: The proportion of true positive predictions among all actual positives.
F1 Score: The harmonic mean of precision and recall, balancing both.

Output:
The results for each classifier are printed, showing their performance on the test set. For example, Logistic Regression achieves the highest accuracy of 65.91%, with an F1 score of 65.19%.

In [8]:
from sklearn.linear_model import LogisticRegression

# Define and train the Logistic Regression model
logistic_model = make_pipeline(StandardScaler(), LogisticRegression())
logistic_model.fit(X_train_imputed, y_train)

# Predict on the test data using the trained Logistic Regression model
predictions_lr = logistic_model.predict(X_test_imputed)

# Create a DataFrame for predictions
predictions_lr_df = pd.DataFrame({'testID': np.arange(1, len(predictions_lr) + 1), 'Status': predictions_lr})

# Save predictions to a CSV file
predictions_lr_df.to_csv("predictions-logistic-regression-test.csv", index=False)


Model Training and Evaluation:

Various classifiers were trained and evaluated, including Decision Tree, Random Forest, SVM, Logistic Regression, and KNN.
Logistic Regression achieved the highest accuracy (65.91%) among the evaluated models, indicating it was the best-performing model on this dataset.

Final Model:
Logistic Regression was selected for further predictions. The model was trained on the preprocessed training data using a pipeline that included feature scaling.
Predictions were made on the test data and saved to a CSV file.

In [9]:
from sklearn.inspection import permutation_importance

# Calculate permutation importance for the Logistic Regression model
result_lr = permutation_importance(logistic_model, X_test_imputed, y_test, n_repeats=10, random_state=42)

# Create a DataFrame for feature importances
feature_importance_perm_lr = pd.DataFrame({'Feature': X_test_imputed.columns, 'Importance': result_lr.importances_mean})

# Sort feature importances in descending order
feature_importance_perm_lr = feature_importance_perm_lr.sort_values(by='Importance', ascending=False)

# Print permutation feature importance for Logistic Regression
print("Permutation Feature Importance for Logistic Regression:")
print(feature_importance_perm_lr)


Permutation Feature Importance for Logistic Regression:
                      Feature  Importance
1              num__Bilirubin    0.064773
0                    num__Age    0.039773
4                 num__Copper    0.028409
16             cat__Ascites_Y    0.020455
15             cat__Ascites_N    0.020455
9            num__Prothrombin    0.019318
18        cat__Hepatomegaly_Y    0.012500
17        cat__Hepatomegaly_N    0.012500
23               cat__Edema_Y    0.012500
10                 num__Stage    0.007955
2            num__Cholesterol    0.004545
7          num__Tryglicerides    0.002273
3                num__Albumin    0.002273
8              num__Platelets    0.001136
13                 cat__Sex_F    0.001136
14                 cat__Sex_M    0.001136
11  cat__Drug_D-penicillamine    0.000000
5               num__Alk_Phos    0.000000
21               cat__Edema_N    0.000000
12          cat__Drug_Placebo    0.000000
6                   num__SGOT   -0.001136
19             cat__

key Predictors: The most important features for predicting the target variable are bilirubin, age, and copper levels.

Less Relevant Features: Features like SGOT, presence of spiders, and certain drug types have little to no positive impact on model performance.

Model Refinement: Based on these results, further model refinement could involve focusing on the high and moderately important features and possibly excluding or re-evaluating the less relevant ones to enhance model efficiency and performance.
